### PR TITLE
DTSPO-14865-replace-kured-v2

### PR DESCRIPTION
**Pull request by JoshuaGlassKainos**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/DTSPO-14865

### Change description ###
added dynamic block to add node os maintenance window to sbox only 

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: 12-kubernetes-cluster.tf
Comment (reespozzi) [inline 12-kubernetes-cluster.tf:140]: Please don't hardcode values this way in a module, it means all envs in the future would need this exact setup if they toggle `enable_node_os_channel_upgrade_nodeimage`

Can you change these to be variables please?

**Review by reespozzi — CHANGES_REQUESTED** [review]

File: 12-kubernetes-cluster.tf
Comment (JoshuaGlassKainos) [inline 12-kubernetes-cluster.tf:140]: need a second opinion on this one, if we lock down freq. to daily only we can remove `day_of_week` not sure to keep it in or not for now.

File: 12-kubernetes-cluster.tf
Comment (reespozzi) [inline 12-kubernetes-cluster.tf:129]:
```suggestion
      condition     = var.node_os_maintenance_window_duration != null && var.node_os_maintenance_window_duration >= 4
```

Haven't read into what this value can be, is it possible to have non-integer? Will need to change if so

File: 12-kubernetes-cluster.tf
Comment (reespozzi) [reply 12-kubernetes-cluster.tf:140]: How does this behave with everything defaulting to null?

File: 12-kubernetes-cluster.tf
Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf:129]: In the docs relating to this they have it as an integer and whole numbers in az cli and json config files, can run a test to see if it can accept string or decimal.

File: 12-kubernetes-cluster.tf
Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf:140]: will need to plan out how it will behave, will also try and implement dynamic block to lock it down to sbox only.

File: 12-kubernetes-cluster.tf
Comment (reespozzi) [reply 12-kubernetes-cluster.tf:129]: Easier to define it as >=4 though right? Easier to read quickly in the code too

Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf:129]: yeah I see what you mean will make that change too

File: 12-kubernetes-cluster.tf
Comment (JoshuaGlassKainos) [reply 12-kubernetes-cluster.tf:140]: back to the drawing board, test failed in a different env so will have another look at it and get your thoughts

File: 10-kubernetes-inputs.tf
Comment (MartyFox) [inline 10-kubernetes-inputs.tf:119]: Match this to the `node_os_maintenance_window_day_of_week` setup, `contains` is expandable versus an OR statement.

Comment (JoshuaGlassKainos) [reply 10-kubernetes-inputs.tf]: Only seeing this now, will try and implement that. thanks Marty

File: 10-kubernetes-inputs.tf
Comment (reespozzi) [inline 10-kubernetes-inputs.tf:132]:
```suggestion
  }
  
```

File: 10-kubernetes-inputs.tf
Comment (reespozzi) [inline 10-kubernetes-inputs.tf:134]:
```suggestion
    condition     = var.node_os_maintenance_window_config.frequency == "Daily" && try(contains(["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"], var.node_os_maintenance_window_config.day_of_week), false)
```

Something like this makes a little more sense for me to read, what do you think?

Comment (JoshuaGlassKainos) [reply 10-kubernetes-inputs.tf:134]: Yeah I agree it simplifies the validation bit better.

File: 10-kubernetes-inputs.tf
Comment (reespozzi) [inline 10-kubernetes-inputs.tf:144]: Again a little confused by this, 
```suggestion
  validation {
    condition     = var.node_os_maintenance_window_config.frequency == "Daily" && var.node_os_maintenance_window_config.day_of_week != null
    error_message = "Invalid day_of_week must not be set for 'Daily' frequency."
  }
}
```

I'm also a little confused what this one is trying to achieve

File: 10-kubernetes-inputs.tf
Comment (reespozzi) [reply 10-kubernetes-inputs.tf]: Oops, meant to keep it as weekly sorry

File: 10-kubernetes-inputs.tf
Comment (JoshuaGlassKainos) [reply 10-kubernetes-inputs.tf]: will give that a shot see how it interacts with the obj

Comment (JoshuaGlassKainos) [reply 10-kubernetes-inputs.tf]: errors out due to the default having 7 attributes vs 6 in the obj, will revert to using the array in the condition

File: 10-kubernetes-inputs.tf
Comment (reespozzi) [inline 10-kubernetes-inputs.tf:141]:
```suggestion
      try(contains(["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"], var.node_os_maintenance_window_config.day_of_week), false))
      error_message = var.node_os_maintenance_window_config.frequency == "Daily" ? "'day_of_week' must not be set for 'Daily' frequency." : "Invalid 'day_of_week', please choose a day of the week."
```

All inline would be nice if it's not going to error

**Review by reespozzi — APPROVED** [review]
LGTM provided tests all OK still

**Comment by JoshuaGlassKainos** [comment]
Tested all types of inputs and all validated as intended. will wait until Marty reviews before merge.

**Comment by JoshuaGlassKainos** [comment]
spoke with Marty and got some suggested changes will implement and review once more

**Review by MartyFox — APPROVED** [review]

**Pull request merged**
